src/Aspire.Hosting/ApplicationModel/ReferenceExpression.cs

```diff
@@ -180,6 +180,18 @@ public void AppendFormatted<T>(T valueProvider) where T : IValueProvider, IManif
             AppendFormatted(valueProvider, format: null);
         }
 
+        /// <summary>
+        /// Appends the formatted value provided by the specified reference expression to the output.
+        /// </summary>
+        /// <param name="valueProvider">A reference expression that supplies the value to be formatted and appended.</param>
+        /// <param name="format">A composite format string that specifies how the value should be formatted, or null to use the default format.</param>
+        /// <remarks>This method is marked obsolete only to prevent usages of this type explicitly.</remarks>
+        [Obsolete("ReferenceExpression instances can't be used in interpolated string with a custom format. Duplicate the inner expression in-place.", error: true)]
+        public void AppendFormatted(ReferenceExpression valueProvider, string format)
```

Comment:
If someone wrote an invalid recursive reference expression then it will hook to this overload and fail the build.